feat: add service accounts options to sso (#852)
defenseunicorns/uds-core: docs/configuration/uds-operator.md

      clientId: httpbin-api-client
      standardFlowEnabled: false
      serviceAccountsEnabled: true

      # By default, Keycloak will not set the audience `aud` claim for service account access token JWTs.
      # You can optionally add a protocolMapper to set the audience.
      # If you map the audience to the same client used for authservice, you can enable access to authservice protected apps with a service account JWT.
      protocolMappers:
        - name: audience
          protocol: "openid-connect"
          protocolMapper: "oidc-audience-mapper"
          config:
            included.client.audience: "uds-core-httpbin" # Set this to match the app's authservice client id
            access.token.claim: "true"
            introspection.token.claim: "true"
            id.token.claim: "false"
            lightweight.claim: "false"
            userinfo.token.claim: "false"
```
Setting `serviceAccountsEnabled: true` requires `standardFlowEnabled: false` and is incompatible with `publicClient: true`.

If needed, multiple clients can be added to the same application: an AuthService client, a device flow client, and as many service account clients as required.

> 

Author: ```suggestion

A keycloak service account JWT can be distinguished by a username prefix of `service-account-` and a new claim called `client_id`.  Note that the `aud` field is not set by default, hence the mapper in the example.
```

      clientId: httpbin-api-client
      standardFlowEnabled: false
      serviceAccountsEnabled: true

      # By default, Keycloak will not set the audience `aud` claim for service account access token JWTs.
      # You can optionally add a protocolMapper to set the audience.
      # If you map the audience to the same client used for authservice, you can enable access to authservice protected apps with a service account JWT.
      protocolMappers:
        - name: audience
          protocol: "openid-connect"
          protocolMapper: "oidc-audience-mapper"
          config:
            included.client.audience: "uds-core-httpbin" # Set this to match the app's authservice client id
            access.token.claim: "true"
            introspection.token.claim: "true"
            id.token.claim: "false"
            lightweight.claim: "false"
            userinfo.token.claim: "false"
```
Setting `serviceAccountsEnabled: true` requires `standardFlowEnabled: false` and is incompatible with `publicClient: true`.

If needed, multiple clients can be added to the same application: an AuthService client, a device flow client, and as many service account clients as required.

> A keycloak service account JWT can be distinguished by a username prefix of `service-account-` and a new claim called `client_id`.  Note that the `aud` field is not set by default, hence the mapper in the example.

Reviewer: ```suggestion
A keycloak service account JWT can be distinguished by a username prefix of `service-account-` and a new claim called `client_id`.  Note that the `aud` field is not set by default, hence the mapper in the example.

```

```yaml
apiVersion: uds.dev/v1alpha1
kind: Package
metadata:
  name: client-cred
  namespace: argo
spec:
  sso:
    - name: httpbin-api-client
      clientId: httpbin-api-client
      standardFlowEnabled: false
      serviceAccountsEnabled: true

      # By default, Keycloak will not set the audience `aud` claim for service account access token JWTs.
      # You can optionally add a protocolMapper to set the audience.
      # If you map the audience to the same client used for authservice, you can enable access to authservice protected apps with a service account JWT.
      protocolMappers:
        - name: audience
          protocol: "openid-connect"
          protocolMapper: "oidc-audience-mapper"
          config:

>             included.client.audience: "uds-core-httpbin" # Set this to match the app's authservice client id

Reviewer: This is the one part of this setup that slightly worries me and I want to make sure I understand. By adding this protocol mapper to a `serviceAccountsEnabled` client, someone can effectively set the audience here to any client. Our current istio/authservice protections would not identify JWTs from this client as being any different from the "real" authservice client since we effectively just check the issuer and audience (which would be 1:1 with the authservice client). I think that is the intent here of course, effectively enabling machine auth on authservice protected apps.

Are there any concerns with this though, especially thinking about multi-tenant clusters? Tenant A could have an authservice protected app, and Tenant B could use a `serviceAccountsEnabled` client to access that app (I think?). Are there things we could/should recommend that the application do to prevent this? Or modifications we could make to our Istio resources to conditionally allow this type of behavior, only if the app is expecting machine auth?

I may also just be missing things here but want to make sure I understand the intent with this flow and any risks/mitigations that are or could be present.

Author: What you're describing is exactly the use-case.  I want to use the `client_id` (or a group / role etc associated with the client) as an authZ key to solve a multi-tenancy issue.  BUT, all of the different tenants need to use the same API.  I originally had this set up so the API did the JWT verification and would accept multiple "audiences" but we always had a goal of getting authservice to take over this role.

We looked at manually supplying the `AuthorizationPolicy` and `RequestAuthentication` CRs in the chart for my project to allow multiple audiences, but figured it's probably safer to let the operator handle this.

I think this is the most intuitive way:  The API is the "audience", which makes sense, but the machines that access the API are doing so on behalf of different companies and development teams, hence the different clients that have different permissions.

Reviewer 2: > Are there any concerns with this though, especially thinking about multi-tenant clusters?

We already don't do any real multi tenant isolation at configuration time, the person deploying a Package CR is basically assumed to be cluster admin I think? Or is that the place of Exemptions, only those are privileged configuration? We already don't isolate Keycloak clients: it turns out two Package CRs can reference the same Keycloak client, and pepr will fetch it out of the store. You can also do other things to hurt multi-tenant isolation, like add any NetworkPolicy you want.

We could consider each custom protocolMappers to be an exemption and require documentation? This is really why I was recommending an allowlist from the beginning for this feature, with the right config you can make your client do anything (probably even takeover Keycloak by overriding claims).

(I really would like to sketch out a threat model and what implied permissions are associated with each CR)

I'm doing a little bit more testing, but as far as I can tell a Keycloak service account JWT can be distinguished by a username prefix of `service-account-` and a new claim called `client_id`. I also think this audience mapper should work with Device Flow, is that maybe worth documenting too?

Author: Also, I think this isn't intuitive, but a `serviceAccount` type client does not automatically come with an `audience` field in keycloak.  Since a lot of people use `aud` to do validation we needed to test a code pattern to add the field.

Reviewer: Yeah sorry @ldgriswold was tracking that the intent of these clients was exactly as described - authenticating using one client against a different authservice app (one that is protected by a different client). My concern was more around potential for two app teams to be working in the same cluster and one to maliciously gain access to the others application using a service account client. Talked async with @bburky about this a bit and I think it's probably not worth addressing at this point since there are other multi-tenant considerations like this that we haven't addressed and don't have a great story around how we want this to look. For now I think we can make the assumption that `Package` CR resources are likely managed by a cluster administrator and/or created via "approved" UDS packages that have been vetted. There might be future things to evaluate with this like restrictions around the audience protocol mapper or service account clients to better ensure they are "cluster admin" functions.

I do think it may be worth adding a quick note around what @bburky commented above...
> a Keycloak service account JWT can be distinguished by a username prefix of `service-account-` and a new claim called `client_id`

Not sure we need to document the device flow + service account piece at this point though.

            access.token.claim: "true"
            introspection.token.claim: "true"
            id.token.claim: "false"
            lightweight.claim: "false"
            userinfo.token.claim: "false"

>       # Optionally, store the client secret a Kubernetes secret

Reviewer: ```suggestion
      # Optionally, store the client secret as a Kubernetes secret
```

I think with our current operator code the secret is actually non-optional, although customizing the output of that secret is optional. Also more generally not following why this would be optional - wouldn't you necessarily want the client secret saved off for use? Might be missing how this client is used though.

Reviewer 2: We have made it required that this client will be `publicClient: false` (has a secret), but this comment refers to `secretName` and `secretTemplate`. As far as I can tell `secretName` is optional but will default to `secretName` if unset.

Author: It would make sense to me if it wasn't optional.  One use case where it might not matter as much is if the client is outside the cluster.  Then you can find a means to pass them the client secret and either grab it from a kubernetes secret or as the keycloak admin.

Reviewer: Yeah I guess to clarify my comment on the comment 😄: The kubernetes secret will always be created unless the client is a public client. So regardless of whether secret name and template are specified here, a k8s secret will be created - it will just end up with a default name and format if you don't provide these (ref code [here](https://github.com/defenseunicorns/uds-core/blob/main/src/pepr/operator/controllers/keycloak/client-sync.ts#L186-L205)). Not to get hung up on one line here in a docs comment but "store the client secret as a Kubernetes secret" is not really optional/conditional on setting these fields was my main point. Comment might be better stated as "optionally set the name and template for the generated kubernetes secret"?

> 
>       # Optionally, store the client secret as a Kubernetes secret
>       secretName: httpbin-api-client
>       secretTemplate:
>         clientId: "clientField(clientId)"
>         clientSecret: "clientField(secret)"

Reviewer: Follow up to the previous thread and suggestion: https://github.com/defenseunicorns/uds-core/pull/852#discussion_r1792162389

Two paths I think we could take here:
1. Just delete this since we capture the SSO secret templating separately [here](https://github.com/defenseunicorns/uds-core/blob/main/docs/configuration/uds-operator.md#example-uds-package-cr-with-sso-templating). I don't _think_ we need to call it out here again?
2. Update the comment to be a bit more accurate:
```suggestion

      # Optionally, provide a custom secret name and template for the secret contents
      secretName: httpbin-api-client
      secretTemplate:
        clientId: "clientField(clientId)"
        clientSecret: "clientField(secret)"
```

Author: If it's captured elsewhere it may make more sense to delete it so we're not crossing the streams too much?

```
Setting `serviceAccountsEnabled: true` requires `standardFlowEnabled: false` and is incompatible with `publicClient: true`.

> If needed, multiple clients can be added to the same application: an authservice client, an device flow client and as many service account clients as required.

Reviewer: ```suggestion
If needed, multiple clients can be added to the same application: an AuthService client, a device flow client, and as many service account clients as required.
```
